pyqtgraph/parametertree/parameterTypes/basetypes.py

```diff
@@ -0,0 +1,421 @@
+        """Called when "add new" button is clicked
+        The parameter MUST have an 'addNew' method defined.
+        """
+        self.param.addNew()
+
+    def addChanged(self):
+        """Called when "add new" combo is changed
+        The parameter MUST have an 'addNew' method defined.
+        """
+        if self.addWidget.currentIndex() == 0:
+            return
+        typ = str(self.addWidget.currentText())
+        self.param.addNew(typ)
+        self.addWidget.setCurrentIndex(0)
+
+    def treeWidgetChanged(self):
+        ParameterItem.treeWidgetChanged(self)
+        tw = self.treeWidget()
+        if tw is None:
+            return
+        self.setFirstColumnSpanned(True)
+        if self.addItem is not None:
+            tw.setItemWidget(self.addItem, 0, self.addWidgetBox)
+            self.addItem.setFirstColumnSpanned(True)
+
+    def addChild(self, child):  ## make sure added childs are actually inserted before add btn
+        if self.addItem is not None:
+            ParameterItem.insertChild(self, self.childCount()-1, child)
+        else:
+            ParameterItem.addChild(self, child)
+
+    def optsChanged(self, param, opts):
+        ParameterItem.optsChanged(self, param, opts)
+
+        if 'addList' in opts:
+            self.updateAddList()
+
+        if hasattr(self, 'addWidget'):
+            if 'enabled' in opts:
+                self.addWidget.setEnabled(opts['enabled'])
+
+            if 'tip' in opts:
+                self.addWidget.setToolTip(opts['tip'])
+
+    def updateAddList(self):
+        self.addWidget.blockSignals(True)
+        try:
+            self.addWidget.clear()
+            self.addWidget.addItem(self.param.opts['addText'])
+            for t in self.param.opts['addList']:
+                self.addWidget.addItem(t)
+        finally:
+            self.addWidget.blockSignals(False)
+
+
+class GroupParameter(Parameter):
+    """
+    Group parameters are used mainly as a generic parent item that holds (and groups!) a set
+    of child parameters.
+
+    It also provides a simple mechanism for displaying a button or combo
+    that can be used to add new parameters to the group. To enable this, the group
+    must be initialized with the 'addText' option (the text will be displayed on
+    a button which, when clicked, will cause addNew() to be called). If the 'addList'
+    option is specified as well, then a dropdown-list of addable items will be displayed
+    instead of a button.
+    """
+    itemClass = GroupParameterItem
+
+    sigAddNew = QtCore.Signal(object, object)  # self, type
+
+    def addNew(self, typ=None):
+        """
+        This method is called when the user has requested to add a new item to the group.
+        By default, it emits ``sigAddNew(self, typ)``.
+        """
+        self.sigAddNew.emit(self, typ)
+
+    def setAddList(self, vals):
+        """Change the list of options available for the user to add to the group."""
+        self.setOpts(addList=vals)
```

Comment:
nit-pick; looks like this file needs a newline at the end.

Reply:
I didn't realize this was a requirement, apologies. I'll configure my settings accordingly

Reply:
I ran a script which found a few other repo files without newlines. Should a quick PR be added for them?

```
examples/exampleLoaderTemplate_pyside2.py
examples/exampleLoaderTemplate_pyside6.py
examples/initExample.py
examples/optics/__init__.py
examples/verlet_chain/__init__.py
pyqtgraph/canvas/__init__.py
pyqtgraph/console/__init__.py
pyqtgraph/dockarea/__init__.py
pyqtgraph/flowchart/__init__.py
pyqtgraph/imageview/__init__.py
pyqtgraph/ordereddict.py
pyqtgraph/util/get_resolution.py
pyqtgraph/util/pil_fix.py
pyqtgraph/widgets/FeedbackButton.py
pyqtgraph/widgets/VerticalLabel.py
tests/test_Point.py
```